homeassistant/components/eheimdigital/light.py

```diff
@@ -0,0 +1,157 @@
+"""EHEIM Digital lights."""
+
+from typing import Any
+
+from eheimdigital.classic_led_ctrl import EheimDigitalClassicLEDControl
+from eheimdigital.types import LightMode
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_EFFECT,
+    EFFECT_OFF,
+    ColorMode,
+    LightEntity,
+    LightEntityFeature,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.color import brightness_to_value, value_to_brightness
+
+from . import EheimDigitalConfigEntry
+from .const import EFFECT_DAYCL_MODE, EFFECT_TO_LIGHT_MODE
+from .coordinator import EheimDigitalUpdateCoordinator
+from .entity import EheimDigitalEntity
+
+BRIGHTNESS_SCALE = (1, 100)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: EheimDigitalConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up EHEIM Digital lights based on a config entry."""
+    coordinator = entry.runtime_data
+
+    known_devices: set[str] = set()
+
+    def check_devices() -> None:
+        if len(coordinator.hub.devices) == 0:
+            return
+
+        registry = er.async_get(hass)
+        existing_entities = er.async_entries_for_config_entry(registry, entry.entry_id)
+
+        entities: list[EheimDigitalClassicLEDControlLight] = []
+
+        new_classic_led_ctrl_devices = [
+            device
+            for device in coordinator.hub.devices.values()
+            if isinstance(device, EheimDigitalClassicLEDControl)
+            and device.mac_address not in known_devices
+        ]
+
+        known_classic_led_ctrl_devices = [
+            device
+            for device in coordinator.hub.devices.values()
+            if isinstance(device, EheimDigitalClassicLEDControl)
+            and device.mac_address in known_devices
+        ]
+
```

Comment:
I think the easiest is to make `known_devices` a list of all devices that you have setup. Even though a device does not have a light, it has "completed" setup, this way you can just use the `set` to easily subtract identifiers and see what needs to be done. (FYI, this code is run every update. We don't have to optimize this to the max, but ideally we keep it as simple as possible)

Reply:
What I think would be ideal is if it was possible to have a "device added" callback which is called by the lib when a new device is found and then the coordinator could forward it to the platforms to set up the entities. This would allow to have devices seen when they are added to the network, but not needing to have it check at each data update.

Reply:
So, I now revamped everything :see_no_evil: It now works completely with callbacks from the lib when new devices are found